### PR TITLE
[Support] Extend ExtensibleRTTI utility to support basic multiple inheritance.

### PR DESCRIPTION
**Pull request by lhames**
Clients can now pass multiple parent classes to RTTIExtends. Each parent class will be inherited from publicly (and non-virtually). The isa, cast, and dyn_cast methods will now work as expected for types with multiple inheritance.

### CODE REVIEW

**Comment by lhames** [comment]
The motivating use-case is ORC memory management: We want `JITLinkMemoryManager`s to be able to inherit from additional interfaces (e.g. for allocating contiguous address ranges for each JITDylib), and to be able to check whether a given memory manager supports a given extended interface.

File: llvm/include/llvm/Support/ExtensibleRTTI.h [hunk 1]
```diff
@@ -110,21 +106,36 @@ class RTTIRoot {
...
 /// };
 ///
-template <typename ThisT, typename ParentT>
-class RTTIExtends : public ParentT {
+template <typename ThisT, typename... ParentTs>
+class RTTIExtends : public ParentTs... {
 public:
   // Inherit constructors from ParentT.
-  using ParentT::ParentT;
+  using ParentTs::ParentTs...;
```
Comment (dwblaikie) [inline llvm/include/llvm/Support/ExtensibleRTTI.h:113]: Does this still apply - I guess it only works if you want to call one of the base ctors and all the other bases support default construction?

Probably OK as a first step - but maybe needs a FIXME to add something like std::pair's piecewise construct (though interesting that tuple doesn't have that... )

**Review by dwblaikie — APPROVED** [review]
Looks good - one optional/commentable thing with the ctor.

File: llvm/include/llvm/Support/ExtensibleRTTI.h [hunk 2]
```diff
@@ -110,21 +106,36 @@ class RTTIRoot {
...
   static const void *classID() { return &ThisT::ID; }
 
   const void *dynamicClassID() const override { return &ThisT::ID; }
 
+  /// Check whether this instance is a subclass of QueryT.
+  template <typename QueryT> bool isA() const { return isA(QueryT::classID()); }
+
   bool isA(const void *const ClassID) const override {
-    return ClassID == classID() || ParentT::isA(ClassID);
+    return ClassID == classID() || parentsAreA<ParentTs...>(ClassID);
```
Comment (dwblaikie) [inline llvm/include/llvm/Support/ExtensibleRTTI.h:123]: Might be able to skip the `parentsAreA` helper:
```
return ClassId == classId() || (ParentTs::isA(ClassID) || ...);
```
(here's at least a little proof of concept I did: https://godbolt.org/z/vGca7jYbT )

**Review by benlangmuir — APPROVED** [review]

**Review by dwblaikie — APPROVED** [review]
Re:single/first base class gets to expose it's non-defsult ctor and all other bases must have default ctors - sounds ok for a first pass.

**Pull request merged**